### PR TITLE
Fix StructReturn handling: properly mark the clobber, and offset actual rets.

### PR DESCRIPTION
**Pull request by cfallin**
The legalization of `StructReturn` was causing issues in the new call-handling code: the `StructReturn` ret was included in the `SigData` as if it were an actual CLIF-level return value, but it is not.

Prior to using regalloc constraints for return values, we unconditionally included rax (or the architecture's usual return register) as a def, so it would be properly handled as "clobbered" by the regalloc. With the new scheme, we include defs on the call only for CLIF-level outputs. Callees with `StructReturn` args were thus not known to clobber the return-value register, and values might be corrupted.

This PR updates the code to include a `StructReturn` ret as a clobber rather than a returned value in the relevant spots. I observed it causing saves/restores of rax in some CLIF that @bjorn3 provided me, but I was having difficulty minimizing this into a test-case that I would be comfortable including as a precise-output case (including the whole thing verbatim would lock down a bunch of other irrelevant details and cause test-update noise later). If we can find a more minimized example I'm happy to include it as a filetest.

Fixes? #5018 (cc @bjorn3 to confirm?).

### CODE REVIEW

**Comment by cfallin** [comment]
Draft PR because this seems to be causing issues in another call filetest that I'm chasing down now...

File: cranelift/codegen/src/machinst/isle.rs [hunk 1]
```diff
@@ -1212,7 +1212,11 @@ macro_rules! isle_prelude_method_helpers {
             let mut outputs = InstOutput::new();
             let mut retval_insts: crate::machinst::abi::SmallInstVec<_> = smallvec::smallvec![];
             for i in 0..num_rets {
-                let ret = self.lower_ctx.sigs()[abi].get_ret(i);
+                // We take the *last* `num_rets` returns of the sig:
+                // this skips a StructReturn, if any, that is present.
+                let sigdata = &self.lower_ctx.sigs()[abi];
+                let idx = (sigdata.num_rets() - num_rets) + i;
```
Comment (elliottt) [inline cranelift/codegen/src/machinst/isle.rs:1218]: How do you feel about changing the loop bounds to `sigdata.num_rets()-num_rets .. sigdata.num_rets()`?

**Comment by bjorn3** [comment]
I just successfully ran the entire cg_clif test suite. Thanks!

File: cranelift/codegen/src/machinst/isle.rs (hunk 1 above)
Comment (cfallin) [reply cranelift/codegen/src/machinst/isle.rs]: Yup, much better!

**Comment by cfallin** [comment]
@elliottt I think this is ready for a proper review now -- thanks!

File: cranelift/codegen/src/machinst/isle.rs [hunk 2]
```diff
@@ -1206,13 +1206,19 @@ macro_rules! isle_prelude_method_helpers {
...
             let mut outputs = InstOutput::new();
             let mut retval_insts: crate::machinst::abi::SmallInstVec<_> = smallvec::smallvec![];
-            for i in 0..num_rets {
-                let ret = self.lower_ctx.sigs()[abi].get_ret(i);
+            // We take the *last* `num_rets` returns of the sig:
+            // this skips a StructReturn, if any, that is present.
+            let sigdata = &self.lower_ctx.sigs()[abi];
+            debug_assert!(num_rets <= sigdata.num_rets());
+            for i in (sigdata.num_rets() - num_rets)..sigdata.num_rets() {
+                let sigdata = &self.lower_ctx.sigs()[abi];
```
Comment (elliottt) [inline cranelift/codegen/src/machinst/isle.rs:1220]: Did this have to be named again to avoid holding a reference to self through the body of the loop?

**Review by elliottt — APPROVED** [review]
Looks good to me!

File: cranelift/codegen/src/machinst/isle.rs (hunk 2 above)
Comment (cfallin) [reply cranelift/codegen/src/machinst/isle.rs:1220]: Yep, exactly, this was a borrow-checker appeasement measure. I'll add a comment to that effect.

**Comment by cfallin** [comment]
I'll backport to the 2.0 release branch once this merges.

**Comment by cfallin** [comment]
@abrown it looks like the `wasi-nn` CI job is hitting 404s when trying to install OpenVino. Could you take a look?

`wasi-nn` support is tier 3, so if is still an issue tomorrow sometime, I'll plan to disable the CI job to get this merged, since it is a miscompilation and needs to not be held back for too long.

**Comment by cfallin** [comment]
> wasi-nn support is tier 3, so if is still an issue tomorrow sometime, I'll plan to disable the CI job to get this merged, since it is a miscompilation and needs to not be held back for too long.

Actually I restarted the job a few times and it looks like a consistent failure from a broken upstream package repository, so I created #5028 to let this be fixed off the critical path...

**Pull request merged**